core/src/main/java/com/linecorp/armeria/server/FallbackService.java

```diff
@@ -56,20 +56,19 @@ private static HttpResponse handleNotFound(ServiceRequestContext ctx, RoutingCon
 
         // Handle the case where '/path' (or '/path?query') doesn't exist
         // but '/path/' (or '/path/?query') exists.
-        final String newPath = oldPath + '/';
-        if (!ctx.config().virtualHost().findServiceConfig(routingCtx.withPath(newPath)).isPresent()) {
+        if (!ctx.config().virtualHost().findServiceConfig(routingCtx.withPath(oldPath + '/')).isPresent()) {
             // No need to send a redirect response because '/path/' (or '/path/?query') does not exist.
             return newHeadersOnlyResponse(HttpStatus.NOT_FOUND);
         }
 
-        // '/path/' (or '/path/?query') exists. Send a redirect response.
-        final String location;
-        if (routingCtx.query() == null) {
-            // '/path/'
-            location = newPath;
-        } else {
-            // '/path/?query'
-            location = newPath + '?' + routingCtx.query();
+        // Use relative path to handle the case where the server is behind a reverse proxy.
+        // The reverse proxy might rewrite the path, so we should use the relative path.
+        // For example, if the proxy rewrite the path /proxy/path -> /path, then we should send the location
+        // with ./path/ so that the client can send the request to /proxy/path/ again.
+        final int index = oldPath.lastIndexOf('/');
+        String location = '.' + oldPath.substring(index) + '/';
```

Comment:
Could this be simply `path/`?

Reply:
Yeah, that also works.
I thought prepending dot was clearer but probably it was a bad idea.
Fixed. Thanks!

Reply:
It's not necessarily a bad idea but I thought it's cheaper and simpler not to add a dot :laughing: